casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -196,14 +208,20 @@ object Running {
         )
     )
 
-  def handleBlockMessage[F[_]: Monad: Log: RequestedBlocks](peer: PeerNode, b: BlockMessage)(
-      casperContains: BlockHash => F[Boolean],
+  def handleBlockMessage[F[_]: Monad: Log: RequestedBlocks: BlockStore](
+      peer: PeerNode,
+      b: BlockMessage
+  )(
+      blockStoreContains: BlockHash => F[Boolean],
       casperAdd: BlockMessage => F[ValidBlockProcessing]
   ): F[Unit] =
-    casperContains(b.blockHash)
+    blockStoreContains(b.blockHash)
       .ifM(
         Log[F].info(s"Received block ${PrettyPrinter.buildString(b.blockHash)} again."),
-        Log[F].info(s"Received ${PrettyPrinter.buildString(b)}.") >> casperAdd(b) >>= (
+        Log[F].info(s"Received ${PrettyPrinter.buildString(b)}.") >> {
+          // TODO check if block is valid here before placing in block store
+          BlockStore[F].put(b)
+        } >> casperAdd(b) >>= (
```

Comment:
This function already abstracts the notion of containing and adding a block. It can also accepts function to add to the store. Or maybe BlockStore.put can be done in `casperAdd` function which has BlockStore already. In both cases BlockStore instance in not necessary.
https://github.com/rchain/rchain/blob/591876bd3872c7e877a580942dfa41d6499eec12/casper/src/main/scala/coop/rchain/casper/engine/Running.scala#L302-L314